Plugin for manual control using sensor_msgs::Joy messages

chataign opened the pull request:
I created a plugin to generate Mavlink manual_control messages from sensor_msgs::Joy input. It is based on the ROS node in px4/src/platforms/ros/nodes/manual_input

Mappings between buttons and the corresponding modes can be configured, as well as the button used to arm/disarm. The plugin has to handle modes & arming because PX4 seems to ignore the buttons flags in the manual_control message (see MavlinkReceiver::handle_message_manual_control) Ideally we would simply forward the button flags via Mavlink and let PX4 handle them.

Maybe this code is too specific, and not generic enough to make sense as a plugin, in which case it shouldn't be merged into master

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+ * @{
+ */
+/*
+ * This file is part of the mavros package and subject to the license terms
```
Please add your copyright string.

vooon commented on mavros_extras/mavros_plugins.xml:
```diff
@@ -23,4 +23,7 @@
         <class name="cam_imu_sync" type="mavplugin::CamIMUSyncPlugin" base_class_type="mavplugin::MavRosPlugin">
 		<description>Publish camera trigger data for synchronisation of IMU and camera frames</description>
 	</class>
+    <class name="joy_control" type="mavplugin::JoyControlPlugin" base_class_type="mavplugin::MavRosPlugin">
```
Looks like 4 spaces instead of tab.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+////////////////////////////////////////////////////////////////////////////////
+
+template<typename Service>
+bool call_service( const string& service_name, const typename Service::Request& req )
```
Why not to use `ros::ServiceClient`?
[tutorial link](http://wiki.ros.org/ROS/Tutorials/WritingServiceClient%28c%2B%2B%29#roscpp_tutorials.2BAC8-Tutorials.2BAC8-WritingServiceClient.Writing_the_Client_Node)

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+		nh_.param("rc_rate_hz",     rc_rate_hz,      10 );
+
+		nh_.param("arming_service", arming_service_, string("/mavros/cmd/arming") );
+		nh_.param("mode_service",   mode_service_,   string("/mavros/set_mode") );
```
Wrong! You should use '~set_mode` etc.

vooon replied:
Also i thing that `nh_.param<std::string>()` looks better.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+		nh_.param("x_axis",         x_.axis,        1   );
+		nh_.param("x_scale",        x_.scale,       1.0 );
+		nh_.param("x_offset",       x_.offset,      0.0 );
+		nh_.param("x_deadzone",     x_.deadzone,    0.2 );
```
Why not move that in subnamespace? E.g. yaml:
```yaml
joy_control:
  arm_button: 2
  x:
    axis: 1
    scale: 1.0
    offset: 0.0
    deadzone: 0.2

  y:
    ...
```

vooon replied:
Also you may place that loaders into `struct Control` to not repeat.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+		nh_.param("r_offset",       r_.offset,      0.0 );
+		nh_.param("r_deadzone",     r_.deadzone,    0.2 );
+
+		joystick_sub_ = nh_.subscribe("/joy", 10, &JoyControlPlugin::handle_joystick, this );
```
Remove `/`.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+		joystick_sub_ = nh_.subscribe("/joy", 10, &JoyControlPlugin::handle_joystick, this );
+		state_sub_ = nh_.subscribe("/mavros/state", 10, &JoyControlPlugin::handle_state, this );
+
+		thread_.reset( new std::thread( &JoyControlPlugin::thread_call, this, rc_rate_hz ) );
```
I'm usually set thread name before storing in `std::thread` (utils).

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+		nh_.param("r_deadzone",     r_.deadzone,    0.2 );
+
+		joystick_sub_ = nh_.subscribe("/joy", 10, &JoyControlPlugin::handle_joystick, this );
+		state_sub_ = nh_.subscribe("/mavros/state", 10, &JoyControlPlugin::handle_state, this );
```
Use `~state`.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+	mavros_msgs::State state_;
+
+	std::mutex msg_mutex_;
+	shared_ptr<std::thread> thread_;
```
Why? `std::thread` has `.swap()` method: https://github.com/mavlink/mavros/blob/master/mavros/include/mavros/setpoint_mixin.h#L73

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+
+	std::mutex msg_mutex_;
+	shared_ptr<std::thread> thread_;
+	mavlink_message_t ctrl_msg_;
```
Why? Use local variable.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+		while (1)
+		{
+			msg_mutex_.lock();
+			UAS_FCU(uas_)->send_message(&ctrl_msg_);
```
Bad idea! At least you abuse drop packet counter buy sending same seq nr.

vooon replied:
Perhaps better do not use thread at all. Joy node has param for minimal rate.

vooon replied:
Also if Joy falls you will send same data and FCU timeout newer occurs.

vooon commented:
Travis uses GCC 4.6: your problem: https://travis-ci.org/mavlink/mavros/jobs/112583245#L4751

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,266 @@
+		// in the sensor_msgs::Joy buttons array
+
+		nh_.param("mode_mapping",   mode_mapping_,  {3,4} );
+		nh_.param("mode_names",     mode_names_,    {"MANUAL","ALTCTL","OFFBOARD"} );
```
Travis falls here. Maybe newer GCC compile that, but not 4.6.

vooon commented:
Also please uncrustify.

chataign commented:
@vooon Thanks for reviewing the code! Much appreciated. I will check your comments today and update the code. After that do you think yourself or someone else would be able to test it?

chataign commented:
@vooon I just committed my changes, I think I addressed all your comments. As you suggested I removed the thread, and I'm now sending the mavlink messages at a fixed rate using a timer. I left the params for the axis/scale/offset/deadzone of each controller as they were, because I find they make it easier to see all default values together. I didn't understand your comment about using yaml for params, if you can give explain I'll have a look again. Thanks!

vooon commented on mavros_extras/mavros_plugins.xml:
```diff
@@ -20,7 +20,10 @@
 		<description>Publish VIBRATION message data from FCU.</description>
 	</class>
-        <class name="cam_imu_sync" type="mavplugin::CamIMUSyncPlugin" base_class_type="mavplugin::MavRosPlugin">
+		<class name="cam_imu_sync" type="mavplugin::CamIMUSyncPlugin" base_class_type="mavplugin::MavRosPlugin">
```
Why? I think that was unintended change.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,237 @@
+		int msg_rate;
+
+		plugin_nh_.param<int>("msg_rate", msg_rate, 10 );
+		plugin_nh_.param<int>("arm_button", arm_button_, 2 );
```
You do not need to specify type in each `NodeHandle::param<>()`, for most calls C++ type out works fine.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,237 @@
+		// in the sensor_msgs::Joy buttons array
+
+		plugin_nh_.param< vector<int> >("mode_mapping", mode_mapping_, {3,4} );
+		plugin_nh_.param< vector<string> >("mode_names", mode_names_, {"MANUAL","ALTCTL","OFFBOARD"} );
```
That was exception. Here you should specify type.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,237 @@
+		set_mode_srv_ = mavros_nh_.serviceClient<mavros_msgs::SetMode>("set_mode");
+
+		arming_srv_.waitForExistence();
+		//set_mode_srv_.waitForExistence();
```
No! You may lock plugin loading! No locks in `initialize()` allowed.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,237 @@
+////////////////////////////////////////////////////////////////////////////////
+
+namespace mavplugin {
+struct Control
```
Inside `class JoyControllPlugin` that ok, but here please add namespace or rename to `JoyControl`.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,237 @@
+#include <mavros_msgs/State.h>
+#include <mavros_msgs/SetMode.h>
+
+using namespace std;
```
I'm usually do not do that. clang says that better list all used classes...

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,237 @@
+	ros::NodeHandle mavros_nh_, plugin_nh_;
+	UAS *uas_;
+	mavros_msgs::State state_;
+	mavlink_message_t ctrl_msg_;
```
That was still there. Just send it in `handle_joystick()` and forget!

vooon commented:
I can't test now. So better if you can check it yourself.

About yaml. That was standard rosparam file storage. I thing that you can reduce code duplication by moving that to `struct Control`. And perhaps it is little more readable.

```cpp
struct JoyAxis {
    void load_params(ros::NodeHandle &priv_nh_, std::string name, size_t axdf) {
        auto nh_ = ros::NodeHandle(priv_nh_, name);

        nh_.param<int>("axis", axis, axdf);
        nh_.param("offset", offset, 0.0);
    }

  size_t axis;
  double offset;
}
...
x_.load_params(pnh, "x", 0);:w
```

vooon commented:
Next thing: you do not need timer at all! Just encode message and send immediately!
FCU has safety timer - it prevents you from uncontrollable flight when your joystick node falls for some reason.

chataign commented:
@vooon I made more changes based on your last comments:
* renamed `Control` struct into `JoyAxis` and moved params loading into structure
* removed timer and sending mavlink message directly in `handle_joystick()`
* no longer waiting for arming/mode services in `initialize()` but warn user if services are not up

I tested the plugin with a Futaba flight controller and a Logitech gamepad. In the end I had to introduce an additional `JoyButton` structure to handle button inputs -- apologies if this change will require more reviewing on your part. This is because on the Futaba controller button states can be represented by multiple flags ([0,0], [0,1], etc), while on the gamepad each button has its own flag.

I found that the best way to configure the plugin for a given controller is to use a yaml file as you suggested. For example my configuration for the logitech gamepad looks like this:

```
arm: { switches: [0], values: [1] }
num_modes: 3
mode1: { switches: [2], values: [0], custom_mode: "MANUAL" }
mode2: { switches: [3], values: [1], custom_mode: "ALTCTL" }
mode3: { switches: [1], values: [1], custom_mode: "OFFBOARD" }
x: { axis: 3, scale: 1000 }
y: { axis: 2, scale: -1000 }
z: { axis: 1, scale: 2000 }
r: { axis: 0, scale: -1000 }
```

And for the futaba controller:

```
arm: { switches: [2], values: [1] }
num_modes: 3
mode1: { switches: [3,4], values: [0,0], custom_mode: "MANUAL" }
mode2: { switches: [3,4], values: [1,0], custom_mode: "ALTCTL" }
mode3: { switches: [3,4], values: [0,1], custom_mode: "OFFBOARD" }
x: { axis: 1, scale: 1000 }
y: { axis: 0, scale: -1000 }
z: { axis: 2, scale: 2000 }
r: { axis: 4, scale: -1000 }
```

then I load the configuration in my launch file using:

```
<rosparam param="joy_control" command="load" file="$(find package)controllers/logitech_gamepad.yaml" />
```

Note that the number of nodes is specified because I don't think it is possible to load arbitrary structures the standard ROS param methods. There are probably ways to manage it but I thought it would be overkill for this plugin.

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,262 @@
+		std::string custom_mode;
+		if ( !button_nh.getParam( "custom_mode", custom_mode ) ) return false;
+
+		return load( nh, button_ns, (Callback) boost::bind( mode_callback, custom_mode ) );
```
Why not `std::bind`?

vooon commented on mavros_extras/src/plugins/joy_control.cpp:
```diff
@@ -0,0 +1,262 @@
+			std::string mode_ns = "mode" + std::to_string(i+1);
+			if ( button.load( plugin_nh_, mode_ns, mode_callback ) ) buttons_.push_back(button);
+			else ROS_ERROR_STREAM("failed to read button for mode=" << mode_ns );
+		}
```
You can request list of param keys and then iterate over it.

vooon commented:
You may request all keys form param server. Here is example [distance_sensor][1].

You should load parameters before node start into `/mavros/joy_control`.
And i'm not sure does parameter cleanup remove all from `/mavros` or not.
Usually for that i use `<fcu>_config.yaml` from `mavros/launch` (or modified copy in robot bringup package).
And i think you should add your default config (same as in plugin code) to that files.

[1]: https://github.com/mavlink/mavros/blob/master/mavros_extras/src/plugins/distance_sensor.cpp#L117

vooon commented:
Oh, and i forgot to ask why not this approach:

1. make simple ManualControl plugin (with single topic)
2. hack `mavteleop` to support it

chataign commented:
@vooon I tried modifying `mavteleop` as you suggested and this is a much better approach. With only a few changes I was able to send RC overrides allowing me to arm, change modes and control the drone. I will test this with a gamepad before I submit my changes for review. Thanks again for your help

chataign commented:
@vooon this pull request should be discarded if you're happy with modifying `mavteleop` instead
